### PR TITLE
Update Helper.php

### PR DESCRIPTION
**Pull request by WJdeBaas**
Remove user-input type-casting.

### Description
Link attributes like "position" should be allowed to be any of the types "integer", "decimal" or "varchar" in custom modules

### Fixed Issues (if relevant)
customization restrictions

### CODE REVIEW

**Comment by aleron75** [comment]
Hello @WJdeBaas, first of all, thank you for your contribution.

>attributes like "position" should be allowed to be any of the types

please, can you provide a use case for types other than `int`? We are just trying to understand under which circumstances we could need a `varchar` position and how it could be used.

Thank you!

File: app/code/Magento/Catalog/Controller/Adminhtml/Product/Initialization/Helper.php [hunk 1]
```diff
@@ -119,6 +124,7 @@ public function __construct(
         \Magento\Catalog\Model\Product\Initialization\Helper\ProductLinks $productLinks,
         \Magento\Backend\Helper\Js $jsHelper,
         \Magento\Framework\Stdlib\DateTime\Filter\Date $dateFilter,
+        Escaper $escaper,
```
Comment (aleron75) [inline app/code/Magento/Catalog/Controller/Adminhtml/Product/Initialization/Helper.php:127]: Let's try to make this change as backward compatible as possible, adding new constructor parameter as **last** and **optional**, with property initialized with object manager if the parameter is not passed.

e.g.

    $this->escaper = $escaper ?: ObjectManager::getInstance()->get(Escaper::class);

**Comment by WJdeBaas** [comment]
Hi Alessandro,

I'll try to explain under which circumstances we could need a varchar
position and how it could be used.

Here's an example: https://www.zandri.nl/catalog/product/view/id/133813

On this page you'll see an exploded view drawing of a scooter carburetor,
where the articles are numbered, but these numbers aren't always typical
integers.
For example: there is 12 en 12-1, and even other drawings can have other
codes like 3a, 3b and 3c for instance.
There are also example where positions are devided with a pipe-symbol (|),
to make it possible to list the same SKU on several different positions in
the drawing.

This functionality has to be migrated to Magento 2, and after some research
and testing I found it's quite easy to reuse and modify parts of the
standard Magento 2 implementations for linked articles like used for
Crosssell and Upsell articles.
But at the moment it's not possible to save divergent values because the
backendcontroller casts the user input, even if the positions' data_type is
declared as a varchar attribute, and the input is saved into the
'catalog_product_link_attribute_varchar' db-table.

In my opinion the variable casting shouldn't be done in the controller, but
in the repository model, where it can check the attributes' type (int,
decimal or varchar), and cast it accordingly just before the data is saved.
I agree with you on the backward compatibility and use the ObjectManager in
case the Escaper class is not passed as a parameter, although the Escaper
class could also be moved the the repository model.

Op do 4 jul. 2019 om 12:35 schreef Alessandro Ronchi <
notifications@github.com>:

**Comment by aleron75** [comment]
Thank you @WJdeBaas for your explanation, it's clear.

My only concern is whether this change should be part of the core or rather should be part of a customized implementation.

Let me check whether we can involve the core team into this discussion.

My best.

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @WJdeBaas thanks for the pull request!

Unfortunately, we can allow only `int` type for link position as this is part of the API: https://github.com/magento/magento2/blob/5588999a108619d09835a2eecbc9579a9e6f27af/app/code/Magento/Catalog/Api/Data/ProductLinkInterface.php#L78

To achieve the desired behaviour you need to extend Magento functionality possibly using an extension attribute for `ProductLinkInterface`

**Comment by WJdeBaas** [comment]
Hi Sergii, Alessandro,

Is it really impossible to apply a minor update to the API to gain more
flexibility  ?
After all, It's "software", not ferroconcrete or titanium ...

    /**
     * Get linked item position
     *
     * @return int|float|string
     */
    public function getPosition();

    /**
     * Set linked item position
     *
     * @param int|float|string $position
     * @return $this
     */
    public function setPosition($position);

Kind regards,
Willem-Jan

Op do 4 jul. 2019 om 15:57 schreef Sergii Ivashchenko <
notifications@github.com>:

**Comment by sidolov** [comment]
@WJdeBaas , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**